node/src/protocol/triple.rs

```diff
@@ -64,19 +80,47 @@ impl TripleManager {
             cait_sith::triples::generate_triple(&self.participants, self.me, self.threshold)
                 .unwrap(),
         );
-        self.generators.insert(id, protocol);
+        self.generators.insert(
+            id,
+            TripleGenerator {
+                protocol,
+                mine: true,
+            },
+        );
     }
 
     /// Take an unspent triple by its id with no way to return it.
     /// It is very important to NOT reuse the same triple twice for two different
     /// protocols.
-    pub fn take(&mut self, id: TripleId) -> Option<TripleGenerationOutput<Secp256k1>> {
+    pub fn take(&mut self, id: TripleId) -> Option<Triple> {
         match self.triples.entry(id) {
             Entry::Vacant(_) => None,
             Entry::Occupied(entry) => Some(entry.remove()),
         }
     }
 
+    /// Take two random unspent triple generated by this node. Either takes both or none.
+    /// It is very important to NOT reuse the same triple twice for two different
+    /// protocols.
+    pub fn take_mine_twice(&mut self) -> Option<(Triple, Triple)> {
+        tracing::info!(mine = ?self.mine, "my triples");
+        if self.mine.len() < 2 {
+            return None;
+        }
+        let id0 = self.mine.pop_front()?;
+        let id1 = self.mine.pop_front()?;
+        tracing::info!(id0, id1, "trying to take two triples");
+        if self.triples.contains_key(&id0) && self.triples.contains_key(&id1) {
+            Some((
+                self.triples.remove(&id0).unwrap(),
+                self.triples.remove(&id1).unwrap(),
+            ))
+        } else {
+            tracing::warn!(id0, id1, "my triples are gone");
+            None
+        }
```

Comment:
This won't have those pesky unwraps:
```suggestion
        if let (Some(triple0), Some(triple1)) = (self.triples.remove(&id0), self.triples.remove(&id1)) {
            Some((triple0, triple1))
        } else {
            tracing::warn!(id0, id1, "my triples are gone");
            None
        }
```

Reply:
wait, nevermind, this isn't the same logic

Reply:
yeah I am still not sure if there is a way to denote this without unwraps